File: CaCompteKit/Sources/Domain/Engine/MatchEngine.swift
import Foundation

public struct MatchEngine: Sendable {
  /// Injectée plutôt qu'un `Date()` implicite (doc 02) : `reduce` reste une fonction pure et
  /// reproductible, y compris rejouée deux fois dans un test.
  private let now: @Sendable () -> Date

  public init(now: @escaping @Sendable () -> Date = Date.init) {
    self.now = now
  }

  /// `(MatchState, MatchEvent) -> MatchState` pure — fondement de l'annulation (on retire
  /// l'événement, on rejoue) et de la synchronisation (on fusionne deux journaux, on rejoue).
  public func reduce(
    _ state: MatchState, _ event: MatchEvent, rules: any GameRules, definition: GameDefinition
  ) -> MatchState {
    var newState = state
    newState.apply(event, rules: rules, definition: definition, occurredAt: now())
    return newState
  }

  /// Dédoublonne par `id`, trie par `(lamport, deviceID)`, replie. Commutatif et idempotent :
  /// `replay(L) == replay(σ(L))` et `replay(L + L) == replay(L)`.
  public func replay(_ log: [StampedEvent], catalog: GameCatalog) throws -> MatchState {
    var seen = Set<UUID>()
    let deduplicated = log.filter { seen.insert($0.id).inserted }
    let sorted = deduplicated.sorted { lhs, rhs in
      lhs.lamport != rhs.lamport ? lhs.lamport < rhs.lamport : lhs.deviceID < rhs.deviceID
    }

    guard let first = sorted.first,
      case .matchCreated(let gameID, let rulesVersion, let variants, let participants) = first.event
    else {
      throw MatchEngineError.missingMatchCreated
    }

    let definition = try catalog.definition(for: gameID, version: rulesVersion)
    let rules = try catalog.rules(for: gameID, version: rulesVersion)

    var state = MatchState(
      matchID: first.id,
      gameID: gameID,
      rulesVersion: rulesVersion,
      variants: variants,
      participants: participants
    )
    for stamped in sorted.dropFirst() {
      state.apply(
        stamped.event, rules: rules, definition: definition, occurredAt: stamped.occurredAt)
    }
    return state
  }
}

public enum MatchEngineError: Error, Sendable, Equatable {
  case missingMatchCreated
}

File: android/domain/src/main/kotlin/com/quimene/domain/engine/MatchEngine.kt
package com.quimene.domain.engine

import com.quimene.domain.model.MatchState
import com.quimene.domain.rules.GameCatalog
import com.quimene.domain.rules.GameDefinition
import com.quimene.domain.rules.GameRules
import java.time.Instant

/** Miroir de `MatchEngineError.swift`. */
sealed class MatchEngineError(
    message: String,
) : Exception(message) {
    data object MissingMatchCreated : MatchEngineError("Le journal ne commence pas par matchCreated")
}

/**
 * Miroir de `MatchEngine.swift`. [now] est une horloge injectée (jamais `Instant.now()` appelé
 * implicitement ailleurs dans le domaine — règle « aucune date implicite », doc 11) pour rendre
 * [reduce] testable de façon déterministe.
 */
class MatchEngine(
    private val now: () -> Instant = Instant::now,
) {
    fun reduce(
        state: MatchState,
        event: MatchEvent,
        rules: GameRules,
        definition: GameDefinition,
    ): MatchState = state.apply(event, rules, definition, occurredAt = now())

    /**
     * Rejeu — dédup par id (idempotent), tri `(lamport, deviceID)` (commutatif quel que soit
     * l'ordre d'arrivée du journal), le premier événement doit être `matchCreated`.
     * `matchID` vient de l'`id` du `StampedEvent`, pas d'un champ explicite de `matchCreated`.
     * `occurredAt` de chaque événement rejoué est **le sien**, jamais [now] — seul le premier
     * état construit (implicitement, à la création) n'a pas cette notion.
     */
    fun replay(
        log: List<StampedEvent>,
        catalog: GameCatalog,
    ): MatchState {
        val seen = HashSet<java.util.UUID>()
        val deduplicated = log.filter { seen.add(it.id) }
        val sorted = deduplicated.sortedWith(compareBy({ it.lamport }, { it.deviceID }))

        val first = sorted.firstOrNull()
        val created =
            (first?.event as? MatchEvent.MatchCreated)
                ?: throw MatchEngineError.MissingMatchCreated

        val definition = catalog.definition(created.gameID, created.rulesVersion)
        val rules = catalog.rules(created.gameID, created.rulesVersion)

        var state =
            MatchState(
                matchID = first.id,
                gameID = created.gameID,
                rulesVersion = created.rulesVersion,
                variants = created.variants,
                participants = created.participants,
            )

        for (stamped in sorted.drop(1)) {
            state = state.apply(stamped.event, rules, definition, occurredAt = stamped.occurredAt)
        }

        return state
    }
}
